File: engine/julia/cliquet_mountain_range_engine.jl
# engine/julia/cliquet_mountain_range_engine.jl
# OptionAlpha Agent — Module W5: Julia Cliquet, Napoleon & Mountain Range Engine

module CliquetMountainRangeEngineJulia

export calculate_lflc_cliquet, calculate_gflc_cliquet, calculate_napoleon, calculate_everest

function calculate_lflc_cliquet(returns::Vector{Float64}, local_floor::Float64, local_cap::Float64)
    return sum(max.(local_floor, min.(returns, local_cap)))
end

function calculate_gflc_cliquet(returns::Vector{Float64}, local_floor::Float64, local_cap::Float64, global_floor::Float64, global_cap::Float64)
    raw = calculate_lflc_cliquet(returns, local_floor, local_cap)
    return max(global_floor, min(global_cap, raw))
end

function calculate_napoleon(returns::Vector{Float64}, max_coupon::Float64)
    worst = minimum(returns)
    return max(0.0, max_coupon + worst)
end

function calculate_everest(returns::Vector{Float64}, coupon::Float64)
    worst = minimum(returns)
    return coupon + worst
end

end # module

File: engine/rust/src/cliquet_mountain_range_engine.rs
// engine/rust/src/cliquet_mountain_range_engine.rs
// OptionAlpha Agent — Module W4: Rust Cliquet, Napoleon & Mountain Range Engine

pub struct CliquetMountainRangeEngineRust;

impl CliquetMountainRangeEngineRust {
    pub fn lflc_cliquet(returns: &[f64], local_floor: f64, local_cap: f64) -> f64 {
        returns.iter().map(|&r| r.max(local_floor).min(local_cap)).sum()
    }

    pub fn gflc_cliquet(returns: &[f64], local_floor: f64, local_cap: f64, global_floor: f64, global_cap: f64) -> f64 {
        let raw = Self::lflc_cliquet(returns, local_floor, local_cap);
        raw.max(global_floor).min(global_cap)
    }

    pub fn napoleon(returns: &[f64], max_coupon: f64) -> f64 {
        let worst = returns.iter().cloned().fold(f64::INFINITY, f64::min);
        (max_coupon + worst).max(0.0)
    }

    pub fn everest(returns: &[f64], coupon: f64) -> f64 {
        let worst = returns.iter().cloned().fold(f64::INFINITY, f64::min);
        coupon + worst
    }
}
